floors/floors.go

```diff
@@ -0,0 +1,89 @@
+package floors
+
+import (
+	"fmt"
+	"math"
+	"math/rand"
+
+	"github.com/mxmCherry/openrtb/v16/openrtb2"
+	"github.com/prebid/prebid-server/currency"
+	"github.com/prebid/prebid-server/openrtb_ext"
+)
+
+const (
+	DEFAULT_DELIMITER      string = "|"
+	CATCH_ALL              string = "*"
+	SKIP_RATE_MIN          int    = 0
+	SKIP_RATE_MAX          int    = 100
+	MODEL_WEIGHT_MAX_VALUE int    = 100
+	MODEL_WEIGHT_MIN_VALUE int    = 0
+	ENFORCE_RATE_MIN       int    = 0
+	ENFORCE_RATE_MAX       int    = 100
+)
+
+// ModifyImpsWithFloors will validate floor rules, based on request and rules prepares various combinations
+// to match with floor rules and selects appripariate floor rule and update imp.bidfloor and imp.bidfloorcur
+func ModifyImpsWithFloors(floorExt *openrtb_ext.PriceFloorRules, request *openrtb2.BidRequest, conversions currency.Conversions) []error {
+	var (
+		floorErrList      []error
+		floorModelErrList []error
+		floorVal          float64
+	)
+	floorData := floorExt.Data
+	if floorData == nil {
+		return nil
+	}
+
+	floorModelErrList = validateFloorSkipRates(floorExt)
+	if len(floorModelErrList) > 0 {
+		return floorModelErrList
+	}
+
+	floorData.ModelGroups, floorModelErrList = selectValidFloorModelGroups(floorData.ModelGroups)
+	if len(floorData.ModelGroups) == 0 {
+		return floorModelErrList
+	} else if len(floorData.ModelGroups) > 1 {
+		floorData.ModelGroups = selectFloorModelGroup(floorData.ModelGroups, rand.Intn)
+	}
+
+	modelGroup := floorData.ModelGroups[0]
+	if modelGroup.Schema.Delimiter == "" {
+		modelGroup.Schema.Delimiter = DEFAULT_DELIMITER
+	}
+
+	floorExt.Skipped = new(bool)
+	if shouldSkipFloors(floorExt.Data.ModelGroups[0].SkipRate, floorExt.Data.SkipRate, floorExt.SkipRate, rand.Intn) {
+		*floorExt.Skipped = true
+		floorData.ModelGroups = nil
```

Comment:
As per section 4.3
Non-selected models are removed when skipRate is false, in case of skipRate is true we would not be either signalling or enforcing floors, hence removed all modelgroups. 
If its required for extra detailed information, we can keep as it in case of skipRate is true